File: frontend/src/api/types/visualizations.ts
import type { AnalysisFilterCondition } from "./analyses";
import type { GraphicalSummaryColumn } from "./analysisResultsExploration";
import type { IndividualsChartResult, RunChartResult } from "./analysisResultsQuality";

export type GraphPreviewType =
  | "box_plot"
  | "individual_value_plot"
  | "histogram"
  | "qq_plot"
  | "ecdf"
  | "scatter_plot"
  | "run_chart"
  | "imr_chart";

export type GraphPreviewLayout = "combined" | "overlay" | "small_multiples";

export interface GraphPreviewRequest {
  dataset_version_id: string;
  filter_snapshot: {
    expression_version: 1;
    conditions: AnalysisFilterCondition[];
  };
  graph_type: GraphPreviewType;
  value_column_ids?: string[];
  x_column_id?: string | null;
  y_column_ids?: string[];
  group_column_id?: string | null;
  order_column_id?: string | null;
  point_limit?: number;
  histogram_bin_count?: number | null;
  layout: GraphPreviewLayout;
}

export interface IndividualValuePreviewPoint {
  series_id: string;
  series_label: string;
  source_column_label: string;
  point_index: number;
  canonical_position: number;
  value: number;
}

export interface IndividualValuePreviewResult {
  point_count: number;
  point_limit: number;
  sampled: false;
  n_total: number;
  n_missing: number;
  n_non_numeric: number;
  points: IndividualValuePreviewPoint[];
}

export interface ScatterPreviewPoint {
  series_id: string;
  series_label: string;
  group: string | null;
  canonical_position: number;
  x: number;
  y: number;
}

export interface ScatterPreviewResult {
  point_count: number;
  point_limit: number;
  sampled: false;
  n_total: number;
  n_excluded: number;
  x_column: GraphPreviewColumnRef;
  y_column: GraphPreviewColumnRef;
  points: ScatterPreviewPoint[];
}

export interface GraphPreviewColumnRef {
  column_id: string;
  display_name: string;
  unit: string | null;
}

interface GraphPreviewPanelBase {
  panel_id: string;
  label: string;
  unit: string | null;
  status: "succeeded" | "failed";
  error_code: string | null;
}

export interface GraphicalSummaryPreviewPanel extends GraphPreviewPanelBase {
  kind: "graphical_summary";
  result: GraphicalSummaryColumn | null;
}

export interface IndividualValuePreviewPanel extends GraphPreviewPanelBase {
  kind: "individual_values";
  result: IndividualValuePreviewResult | null;
}

export interface ScatterPreviewPanel extends GraphPreviewPanelBase {
  kind: "scatter";
  result: ScatterPreviewResult | null;
}

export interface RunChartPreviewPanel extends GraphPreviewPanelBase {
  kind: "run_chart";
  result: RunChartResult | null;
}

export interface ImrChartPreviewPanel extends GraphPreviewPanelBase {
  kind: "imr_chart";
  result: IndividualsChartResult | null;
}

export type GraphPreviewPanel =
  | GraphicalSummaryPreviewPanel
  | IndividualValuePreviewPanel
  | ScatterPreviewPanel
  | RunChartPreviewPanel
  | ImrChartPreviewPanel;

export interface GraphPreviewResponse {
  visualization_schema_version: 1;
  graph_type: GraphPreviewType;
  dataset_version_id: string;
  source_schema_hash: string;
  filter_snapshot_sha256: string;
  preview_config_sha256: string;
  row_count_total: number;
  row_count_included: number;
  warnings: string[];
  layout: GraphPreviewLayout;
  panels: GraphPreviewPanel[];
}

File: backend/app/api/v1/schemas/visualizations.py
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.api.v1.schemas.analyses import AnalysisFilterSnapshot

GraphType = Literal[
    "box_plot",
    "individual_value_plot",
    "histogram",
    "qq_plot",
    "ecdf",
    "scatter_plot",
    "run_chart",
    "imr_chart",
]
GraphLayout = Literal["combined", "overlay", "small_multiples"]
GraphComparisonMode = Literal["multiple_values", "one_value_by_group"]
ScatterMode = Literal["fixed_x_multiple_y", "multiple_x_fixed_y"]


class GraphPreviewRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    dataset_version_id: UUID
    filter_snapshot: AnalysisFilterSnapshot = Field(default_factory=AnalysisFilterSnapshot)
    graph_type: GraphType
    value_column_ids: list[str] = Field(default_factory=list)
    scatter_mode: ScatterMode | None = None
    x_column_ids: list[str] = Field(default_factory=list)
    x_column_id: str | None = None
    y_column_ids: list[str] = Field(default_factory=list)
    group_column_id: str | None = None
    order_column_id: str | None = None
    point_limit: int = Field(default=1000, ge=10, le=2000)
    histogram_bin_count: int | None = Field(default=None, ge=1, le=200)
    layout: GraphLayout = "small_multiples"
    comparison_mode: GraphComparisonMode = "multiple_values"
    group_order_policy: Literal["first_occurrence"] = "first_occurrence"
    missing_group_policy: Literal["exclude"] = "exclude"

    @model_validator(mode="after")
    def validate_roles(self) -> "GraphPreviewRequest":
        if self.graph_type == "scatter_plot":
            if not self.x_column_ids and self.x_column_id is not None:
                self.x_column_ids = [self.x_column_id]
            mode = self.scatter_mode or "fixed_x_multiple_y"
            self.scatter_mode = mode
            if mode == "fixed_x_multiple_y":
                if len(self.x_column_ids) != 1 or not 1 <= len(self.y_column_ids) <= 6:
                    raise ValueError("fixed_x_multiple_y requires one X and 1-6 Y columns")
            elif len(self.y_column_ids) != 1 or not 1 <= len(self.x_column_ids) <= 6:
                raise ValueError("multiple_x_fixed_y requires 1-6 X and one Y column")
            if len(set(self.x_column_ids)) != len(self.x_column_ids):
                raise ValueError("scatter_plot X columns must be unique")
            if len(set(self.y_column_ids)) != len(self.y_column_ids):
                raise ValueError("scatter_plot Y columns must be unique")
            if set(self.x_column_ids) & set(self.y_column_ids):
                raise ValueError("scatter_plot X and Y columns must be different")
            return self

        if not self.value_column_ids:
            raise ValueError(f"{self.graph_type} requires value_column_ids")
        limits = {
            "box_plot": 12,
            "individual_value_plot": 8,
            "histogram": 8,
            "qq_plot": 8,
            "ecdf": 6,
            "run_chart": 6,
            "imr_chart": 6,
        }
        if len(self.value_column_ids) > limits[self.graph_type]:
            raise ValueError(f"{self.graph_type} has too many value columns")
        comparison_graphs = {"box_plot", "individual_value_plot", "imr_chart"}
        if self.comparison_mode == "one_value_by_group":
            if self.graph_type not in comparison_graphs:
                raise ValueError("group comparison is not supported for this graph type")
            if self.group_column_id is None or len(self.value_column_ids) != 1:
                raise ValueError("group comparison requires one value and one group column")
        elif self.group_column_id is not None and self.graph_type in comparison_graphs:
            raise ValueError("multiple value comparison does not accept group_column_id")
        return self


class GraphPreviewPanel(BaseModel):
    model_config = ConfigDict(extra="forbid")

    panel_id: str
    kind: Literal[
        "graphical_summary",
        "individual_values",
        "scatter",
        "run_chart",
        "imr_chart",
    ]
    label: str
    unit: str | None
    status: Literal["succeeded", "failed"]
    error_code: str | None = None
    result: dict[str, Any] | None = None


class GraphPreviewResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    visualization_schema_version: Literal[3] = 3
    graph_type: GraphType
    dataset_version_id: UUID
    source_schema_hash: str
    filter_snapshot_sha256: str
    preview_config_sha256: str
    row_count_total: int
    row_count_included: int
    warnings: list[str]
    layout: GraphLayout
    comparison_mode: GraphComparisonMode
    scatter_mode: ScatterMode | None = None
    group_order_policy: Literal["first_occurrence"]
    missing_group_policy: Literal["exclude"]
    missing_group_row_count: int = Field(ge=0)
    panels: list[GraphPreviewPanel]
